db/compaction_picker.cc

```diff
@@ -42,29 +42,44 @@ uint64_t TotalCompensatedFileSize(const std::vector<FileMetaData*>& files) {
 bool FindIntraL0Compaction(const std::vector<FileMetaData*>& level_files,
                            size_t min_files_to_compact,
                            uint64_t max_compact_bytes_per_del_file,
-                           CompactionInputFiles* comp_inputs) {
-  size_t compact_bytes = static_cast<size_t>(level_files[0]->fd.file_size);
+                           CompactionInputFiles* comp_inputs,
+                           SequenceNumber earliest_mem_seqno) {
+  // Do not pick ingested file when there is at least one memtable not flushed
+  // which of seqno is overlap with the sst.
+  size_t start = 0;
+  for (; start < level_files.size(); ++start) {
+    if (level_files[start]->being_compacted) {
+      return false;
+    }
+    if (level_files[start]->fd.largest_seqno <= earliest_mem_seqno) {
```

Comment:
Note this is `<=` even though our conversation has always implied it would be a `<`. When memtable rolls, RocksDB uses the latest consumed seqnum to initialize the new memtable. That seqnum exactly equals the latest flushed file's largest seqnum. The existing intra-L0 tests uncovered this. I think it was masked in the upstream PR due to this condition being `&&`'d with `being_compacted`.